Add upsells, cross-sells, and related products to Store API Products endpoint

Abdalsalaam opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

<!-- For bug fixes: If known, please provide links to help with traceability and escape analysis. -->
<!-- Please include a link to the issue of the bug being fixed, if one doesn't exist please create it. -->
<!-- If the PR that introduced the bug is known, please also add its link below. -->
This PR adds embeddable `_links` for upsells, cross-sells, and related products to the Store API Products endpoints, along with a new `related` query parameter to filter products by related product ID.

Closes #62601 .

### Added :

1. **Product Links (`_links`)** :
   - `self` - Link to the current product.
   - `collection` - Link to the products collection.
   - `up` - Link to parent product (for variations only).
   - `upsells` - Embeddable link to fetch upsell products (if configured).
   - `cross_sells` - Embeddable link to fetch cross-sell products (if configured).
   - `related` - Embeddable link to fetch related products.

2. **`related` Query Parameter** - Filter products by related product ID:
   ```
   GET /wc/store/v1/products?related=34
   ```
Returns products related to product ID 34 (based on shared categories/tags), excluding the main product itself.

## API Response Example

```json
{
  "id": 15,
  "name": "Product Name",
  "_links": {
    "self": [{"href": ".../products/15"}],
    "collection": [{"href": ".../products"}],
    "upsells": [{"href": ".../products?include=10,20", "embeddable": true}],
    "cross_sells": [{"href": ".../products?include=30", "embeddable": true}],
    "related": [{"href": ".../products?related=15&per_page=10", "embeddable": true}]
  }
}
```

With `?_embed`
WordPress automatically fetches and embeds the linked resources:

```json
{
  "id": 15,
  "name": "Product Name",
  "_links": { ... },
  "_embedded": {
    "upsells": [
      {"id": 10, "name": "Upsell Product 1", ...},
      {"id": 20, "name": "Upsell Product 2", ...}
    ],
    "cross_sells": [
      {"id": 30, "name": "Cross-sell Product", ...}
    ],
    "related": [
      {"id": 5, "name": "Related Product 1", ...},
      ...
    ]
  }
}
```

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. **Verify `_links` in collection response**
   - Request `GET /wc/store/v1/products`.
   - Confirm each product includes `_links` with `self`, `collection`, and `related`.
   - Products with upsells/cross-sells should also include those links.

2. **Verify `_links` for single product**
   - Request `GET /wc/store/v1/products/{id}`
   - Confirm `_links` are present with `embeddable: true` for upsells, cross_sells, and related

3. **Verify `?_embed` returns embedded data**
   - Request `GET /wc/store/v1/products/{id}?_embed`.
   - Confirm `_embedded` object contains full product data for linked resources.

4. **Test `related` query parameter**
   - Create products in the same category.
   - Request `GET /wc/store/v1/products?related={product_id}`.
   - Verify related products are returned and the main product is excluded.

<!-- End testing instructions -->

### Testing that has already taken place:

<!-- Detail any testing that has already been conducted. -->
<!-- Include environment details such as hosting type, plugins, theme, store size, store age, and relevant settings. -->
<!-- Mention any analysis performed, such as assessing potential impacts on environment attributes and other plugins, performance profiling, or LLM/AI-based analysis. -->
<!-- Within the testing details you provide, please ensure that no sensitive information (such as API keys, passwords, user data, etc.) is included in this public pull request. -->

### Milestone

<!-- DO NOT remove or modify this section (other than to check the box). -->

<!-- milestone-target-selection -->
- [x] Automatically assign milestone for the **[next WooCommerce version](../blob/trunk/plugins/woocommerce/woocommerce.php#L6)**
<!-- /milestone-target-selection -->

> **Note:** Check the box above to have the milestone automatically assigned when merged.
> Alternatively (e.g. for point releases), manually assign the appropriate milestone.


### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [x] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [x] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [x] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->
Add embeddable _links for upsells, cross-sells, and related products to Store API Products endpoint.
</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

Abdalsalaam committed:
Add related_ids, upsell_ids, and cross_sell_ids

Abdalsalaam committed:
add tests

Commit:
Add changefile(s) from automation for the following project(s): woocommerce

Abdalsalaam committed:
Update @SInCE version placeholders

senadir commented:
Thank you @Abdalsalaam I'd like to take some time to think about this, ideally I think we should start exploring using embeds for such things instead of including them in the main request, this would mean only the person who needs them is going to get them.

Before we do that ideally I'd like to check the existing response to see what's being inlined to see opportunities for embedding

senadir commented:
I agree, I commented before checking the code and I see it only includes ids not the whole items. I might need to go back and forth on the name as i personally might just do `related_products`, `cross_sells`... Plenty of those things we're changing in V4 of REST API

Abdalsalaam commented:
> I agree, I commented before checking the code and I see it only includes ids not the whole items. I might need to go back and forth on the name as i personally might just do `related_products`, `cross_sells`... Plenty of those things we're changing in V4 of REST API

@senadir Makes sense! I followed the REST API v3 naming (related_ids, upsell_ids, cross_sell_ids) for consistency, but happy to rename to related_products, cross_sells, upsells if that aligns better with the V4 direction.
Let me know your preference and I'll update the PR.

Thanks!

tyxla reviewed (commented):
I'm not 100% convinced we want to add these to the product schema in the Store API. 

FWIW, they are part of the REST API product schema: https://woocommerce.github.io/woocommerce-rest-api-docs/#retrieve-a-product

I'm imagining there's a reason why we're not exposing these. Perhaps we don't want to expose publicly which products we're upselling or cross-seling? 

cc @woocommerce/rubik on this one and specifically @mikejolley who worked heavily on the majority of these endpoints. I didn't find any evidence why we chose not to add those initially.

Abdalsalaam commented:
Thank you @tyxla for clarifying this! 

I see the concern, and it’s a fair point to raise. I’m not sure the reason these fields aren’t currently exposed is about keeping this data private. For example, the Cart API already returns cross-sells when items are added to the cart, so this information is already accessible through the Store API, just in a different context.

From my side, the main value in adding these fields is around supporting mobile apps and headless WooCommerce implementations. At the moment, if you’re building a product page using only the Store API, you need custom solutions just to display related products, upsells, or cross-sells which are pretty standard parts of an e-commerce product page.

Making these fields available would help avoid those workarounds and make the Store API more practical for common use cases out of the box. I’m happy to go with whatever the team decides here. If the outcome is to move forward with adding these fields, I’m more than happy to update the PR accordingly. 😄

tyxla commented on plugins/woocommerce/src/StoreApi/Schemas/V1/ProductSchema.php:
```diff
@@ -760,6 +790,42 @@ function ( $child ) {
+	 * @return array<int>
+	 */
+	protected function get_related_ids( \WC_Product $product ): array {
+		return array_values( array_map( 'absint', wc_get_related_products( $product->get_id() ) ) );
```
If we've got an array of non-negative integers, why do we need to pass it through `array_values()`?

Did you mean to call them in the opposite order?


```suggestion
		return array_map( 'absint', array_values( wc_get_related_products( $product->get_id() ) ) );
```

Abdalsalaam replied:
Yes I meant the opposite! thanks for catching this

tyxla commented on plugins/woocommerce/src/StoreApi/Schemas/V1/ProductSchema.php:
```diff
@@ -760,6 +790,42 @@ function ( $child ) {
+	 * @return array<int>
+	 */
+	protected function get_upsell_ids( \WC_Product $product ): array {
+		return array_values( array_map( 'absint', $product->get_upsell_ids() ) );
```
`get_upsell_ids()` will already return IDs, no? So we shouldn't need to call `array_values` at all:


```suggestion
		return array_map( 'absint', $product->get_upsell_ids() );
```

tyxla commented on plugins/woocommerce/src/StoreApi/Schemas/V1/ProductSchema.php:
```diff
@@ -760,6 +790,42 @@ function ( $child ) {
+	 * @return array<int>
+	 */
+	protected function get_cross_sell_ids( \WC_Product $product ): array {
+		return array_values( array_map( 'absint', $product->get_cross_sell_ids() ) );
```
`get_cross_sell_ids()` will already return IDs, no? So we shouldn't need to call `array_values` at all:


```suggestion
		return array_map( 'absint', $product->get_cross_sell_ids() );
```

tyxla commented on plugins/woocommerce/tests/php/src/Blocks/StoreApi/Routes/Products.php:
```diff
@@ -265,6 +265,105 @@ public function test_without_image() {
+		$main_product->set_category_ids( array( $category['term_id'] ) );
+		$main_product->save();
+
+		delete_transient( 'wc_related_' . $main_product->get_id() );
```
Is this necessary? If yes, should we fix this so saving a product triggers deletion of its related products transient, rather than adding it here to the test?

senadir replied:
It seems it was done on purpose https://github.com/woocommerce/woocommerce/pull/59642

Abdalsalaam replied:
> Is this necessary? If yes, should we fix this so saving a product triggers deletion of its related products transient, rather than adding it here to the test?

No, it's not necessary. The PR @senadir  mentioned added `wc_related_` to `delete_product_specific_transients()`, so it's already cleared on WC_Product::save(). Removed

https://github.com/Abdalsalaam/woocommerce/blob/b1c228c0ea84e518fe418e7ade680b525363a75c/plugins/woocommerce/src/Internal/Utilities/ProductUtil.php#L35

https://github.com/Abdalsalaam/woocommerce/blob/7b78e2df7e7c64aa5d8d703d295139e80654a023/plugins/woocommerce/includes/wc-product-functions.php#L627

tyxla reviewed (commented):
Thanks for working on this one @Abdalsalaam 🙌 

I'll leave the final review to @senadir and the rest of @woocommerce/rubik, but I had a few questions as I was glancing through it.

Abdalsalaam committed:
fix returned array mapping

Abdalsalaam committed:
remove unnecessary 'delete_transient'

Abdalsalaam committed:
add the new fields to product endpoint response doc

tyxla reviewed (commented):
Thanks @Abdalsalaam, my feedback has been addressed.

Can someone from @woocommerce/rubik please have a last look and help land this? Thanks!

mikejolley commented on plugins/woocommerce/src/StoreApi/Schemas/V1/ProductSchema.php:
```diff
@@ -760,6 +790,42 @@ function ( $child ) {
+	 * @return array<int>
+	 */
+	protected function get_upsell_ids( \WC_Product $product ): array {
+		return array_map( 'absint', $product->get_upsell_ids() );
```
I understand this is pretty much copied from the REST API implementation, however, given we should expect IDs as integers here, could we not move this handling into the product class? 

If the forced `absint` was applied within `WC_Product` `set_upsell_ids` and `set_cross_sell_ids` I think we could just use `$product->get_upsell_ids()` directly instead of including these methods. What do you think?

Abdalsalaam replied:
@mikejolley I agree with this! moving the absint handling into the setters would make the code cleaner.

mikejolley reviewed (commented):
Generally I agree with what has been said already; this is not private so its safe to include in the Store API responses.

`upsell_ids` and `cross_sell_ids` are pulled from the current product data so should be fine to include. I have some concerns with including `related_ids` due to performance. 

While I can see this does have a transient cache, should that cache not exist or be invalidated, there would be an expensive lookup query per product returned in order to find what IDs are related to each product. This won't be needed by all consumers.

Unlike WC REST API, I don't believe the Store API allows you to narrow the returned dataset using `_fields` so there is no way to opt out of related_ids in the response.

There is also the practicality aspect of using this data. If you wanted to render related products, you only have the IDs. So in your client you'd need to then do follow up API requests to get product data, which is workable, but I feel like an `_embed` solution like we've been experimenting with in REST API v4 would be superior. 

With that in mind, I would suggest keeping `upsell_ids` and `cross_sell_ids` in this PR, but removing `related_ids`.

What do you think @senadir @opr?

senadir reviewed (dismissed):
With this I think the next change would be to move away from using `x_ids` toward just an embed with `x` label so upsells, cross_sells, and related.

See how we're doing it in V4 Orders endpoint https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/src/Internal/RestApi/Routes/V4/Orders/Controller.php#L202-L250

Basically calling the product route with those ids:

```php
'upsells'         => array(
	'href'       => add_query_arg(
		array( 'include' => implode( ',', $item->get_upsell_ids() )  ),
		rest_url( sprintf( '/%s/products', $this->namespace ) )
	),
	'embeddable' => true,
),
```

Abdalsalaam committed:
revert adding related_ids, upsell_ids, and cross_sell_ids to the response

Abdalsalaam committed:
Add trait ProductLinksTrait
Shared functionality for preparing product links including embeddable links for upsells, cross-sells, and related products

Abdalsalaam committed:
Use ProductLinksTrait to add _links to the items

Abdalsalaam committed:
Use ProductLinksTrait to add _links to the item object

Abdalsalaam committed:
Add 'embed' context

woocommercebot committed:
Add changefile(s) from automation for the following project(s): woocommerce

Abdalsalaam commented:
Thank you @senadir and @mikejolley for this very helpful reviews

I found that `_links` were never being returned in the Store API Products endpoint, the `prepare_links()` method existed but was never called

I've fixed this, but I'm not sure if including _links in the collection endpoint (/products) is necessary or desired. The concern is that `wc_get_related_products()` runs for each product in the list, which could impact performance on large collections.                                                                                                                                                                                                                                          

Options:
1. Keep `_links` on both collection and single product endpoints (current implementation).
2. Only add `_links` to single product endpoints (/products/{id}, /products/{slug}).
3. Add _links to collection but skip the related link (only include upsells and cross_sells which don't require a DB query)

would appreciate feedback on the preferred approach, once confirmed, I'll add tests and documentation

Abdalsalaam committed:
force absint for get_upsell_ids and get_cross_sell_ids

senadir commented on plugins/woocommerce/src/StoreApi/Routes/V1/Products.php:
```diff
@@ -78,7 +81,8 @@ protected function get_route_response( \WP_REST_Request $request ) {
 			foreach ( $query_results['objects'] as $object ) {
-				$data               = rest_ensure_response( $this->schema->get_item_response( $object ) );
+				$data = rest_ensure_response( $this->schema->get_item_response( $object ) );
+				$data->add_links( $this->prepare_links( $object, $request ) );
```
Can't seem to find a case in which wc_get_product return false because we filter them out early on, can you elaborate @coderabbitai

senadir commented:
I seem to be unable to get cross_sells to expand in my testing, either by adding `?embed` or `?embed=cross_sells`, only upsells is expanding. I think in my testing this is due the fact a cross-sell product is not purchasable, in a sense:
- `GET http://localhost:10003/wp-json/wc/store/v1/products/31` works
- `GET http://localhost:10003/wp-json/wc/store/v1/products?include=31` returns an empty object.

This might be a bug at the Store API products endpoint level.

Abdalsalaam committed:
draft filter by related products query

Abdalsalaam committed:
Use related param for related link

Abdalsalaam committed:
Limit to 5 related products

Abdalsalaam committed:
Add Product Links and Embedding to the documentation

Abdalsalaam committed:
fix markdown

Abdalsalaam committed:
Add Products endpoint _links tests

mikejolley commented on plugins/woocommerce/src/StoreApi/Utilities/ProductLinksTrait.php:
```diff
@@ -0,0 +1,72 @@
+			'href'       => add_query_arg(
+				array(
+					'related'  => $item->get_id(),
+					'per_page' => 5,
```
Let's default to 10 like product controller does.

mikejolley commented on plugins/woocommerce/src/StoreApi/Utilities/ProductQuery.php:
```diff
@@ -251,6 +251,32 @@ function ( $value ) {
+		// Filter by related products.
+		if ( ! empty( $request['related'] ) ) {
+			$related_to  = array_map( 'absint', $request['related'] );
+			$limit       = ! empty( $request['per_page'] ) ? (int) $request['per_page'] : -1;
```
Recently product controller capped at 100 to prevent abuse. Let's do this here as well.
```suggestion
			$limit       = ! empty( $request['per_page'] ) ? (int) $request['per_page'] : 100;
```

Abdalsalaam replied:
that make sense !

mikejolley commented on plugins/woocommerce/src/StoreApi/Utilities/ProductQuery.php:
```diff
@@ -251,6 +251,32 @@ function ( $value ) {
+			$limit       = ! empty( $request['per_page'] ) ? (int) $request['per_page'] : -1;
+			$all_related = array();
+
+			foreach ( $related_to as $product_id ) {
```
Just to prevent someone querying too much data, I think we should just allow one related product ID per query. Can we make `related` support only one ID?

Abdalsalaam replied:
@mikejolley Thanks for reviewing this! i pushed changes to address these

mikejolley reviewed (changes requested):
This is looking great now. Just a few small changes to pagination and `related` param behaviour to prevent potential abuse (since the queries can be heavy) and this should be ready.

Thanks again

Abdalsalaam committed:
set per_page to 10

Abdalsalaam committed:
cape per_page limit at 100 to prevent abuse

Abdalsalaam committed:
make related support only one ID

Abdalsalaam committed:
update param

Abdalsalaam committed:
related products doc

Abdalsalaam committed:
update tests for related

Abdalsalaam committed:
update related per_page test

Abdalsalaam committed:
update related link example

Abdalsalaam committed:
test Related product should be returned

Abdalsalaam committed:
remove redundant array( $product_id ) exclusion

Abdalsalaam committed:
Adding validate_callback to ensures consistent validation behavior

Abdalsalaam committed:
lint fixes

mikejolley committed:
Guard is not needed for autoloaded classes

mikejolley committed:
Correctly handle errors from rest_ensure_response

mikejolley committed:
Use prepare_item_for_response in products route

mikejolley committed:
Update phpstan baseline for CI

Abdalsalaam committed:
Added 'embed' to context arrays

Abdalsalaam committed:
schema natively includes 'embed'

mikejolley committed:
baseline update

mikejolley reviewed (approved):
Thanks for fixing those issues @Abdalsalaam 

CI is now green and I've rebased everything. Merging!
